### PR TITLE
feat(example): new example providing hub auth for private app users

### PR DESCRIPTION
**Pull request by andrewxhill**
Signed-off-by: andrew <andrew@textile.io>

### CODE REVIEW

**Review by carsonfarmer — APPROVED** [review]

File: browser-hub-app/src/server/api.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  const payload = {
+    ...auth,
+    key: process.env.USER_API_KEY,
+  };
+  
+  /** Return the auth in a JSON object */
+  ctx.body = payload
+  
+  await next();
+});
```
Comment (andrewxhill) [inline browser-hub-app/src/server/api.ts]: Simple auth, no user validation

File: browser-hub-app/src/server/wss.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+      }
+    } catch (error) {
+      /** Notify our client of any errors */
+      ctx.websocket.send(JSON.stringify({
+        type: 'error',
+        value: error.message,
+      }))
+    }
+  });
+});
```
Comment (andrewxhill) [inline browser-hub-app/src/server/wss.ts]: Full auth, user validation and client-side challenge.

File: browser-hub-app/src/client/index.ts [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    }
+
+    /** Use the identity to request a new API token */
+    this.auth = await loginWithChallenge(this.id);
+
+    console.log('Verified on Textile API')
+    displayStatus();
+
+    /** Store the access control metadata */
+    this.context = Context.fromUserAuth(this.auth, 'http://localhost:3007')
```
Comment (andrewxhill) [inline browser-hub-app/src/client/index.ts]: Login using the complete auth with user validation (see loginWithChallenge method in same file)

File: browser-hub-app/src/client/index.ts [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    /** The simple auth endpoint doesn't provide a user's Hub API Token */
+    const client = new Client(this.context)
+    const token = await client.getToken(this.id)
+
+    /** Update our context, including the token */
+    this.auth = {
+      ...this.auth,
+      token: token,
+    }
+    this.context = Context.fromUserAuth(this.auth)
```
Comment (andrewxhill) [inline browser-hub-app/src/client/index.ts]: simple auth where server just trusts every user

File: browser-hub-app/README.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+## Configure
+
+Create a `.env` file in the root of your project. Ensure you never check this file into your repo or share it, it contains your User Group Key Secret.
+
+```bash
+cp example.env .env
+```
+
+Then replace the `USER_API_KEY` and `USER_API_SECRET` values with those you create using the Textile Hub and your own account or org.
```
Comment (carsonfarmer) [inline browser-hub-app/README.md]:
```suggestion
Then replace the `USER_API_KEY` and `USER_API_SECRET` values with those you create using the Textile Hub and your own account or org (see [docs.textile.io](https://docs.textile.io) for details).
```

File: browser-hub-app/src/server/index.ts [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+app.use( logger() );
+app.use( bodyParser() );
+
+app.use(serve(__dirname + '/../client'));
+
+/**
+ * Start HTTP Routes
+ */
+const router = new Router();
+app.use( router.routes() ).use( router.allowedMethods() );
```
Comment (carsonfarmer) [inline browser-hub-app/src/server/index.ts]: Is this being called twice? Once here and once below? Or does it beed to be called as routes are added?

**Review by carsonfarmer — APPROVED** [review]

File: browser-hub-app/src/server/index.ts (hunk 6 above)
Comment (andrewxhill) [reply browser-hub-app/src/server/index.ts]: i don't see it twice, i see it for different router instances...?

**Pull request merged**